# Module 2 - Mini-project 2

## Libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer

## Concaténation

In [2]:
df1 = pd.read_csv("C:/Users/amandine.gauberville/dataV2-labs/module-2/mini-project2/route-finder-3rd-5.6.csv")
df2 = pd.read_csv("C:/Users/amandine.gauberville/dataV2-labs/module-2/mini-project2/route-finder-rock-5.10-5.12d.csv")
df3 = pd.read_csv("C:/Users/amandine.gauberville/dataV2-labs/module-2/mini-project2/route-finder-rock-5.7-5.10a-b.csv")
df4 = pd.read_csv("C:/Users/amandine.gauberville/dataV2-labs/module-2/mini-project2/route-finder-rock-5.13a-5.14d.csv")

In [3]:
df = pd.concat([df1,df2,df3,df4])

In [4]:
df.columns

Index(['Route', 'Location', 'URL', 'Avg Stars', 'Your Stars', 'Route Type',
       'Rating', 'Pitches', 'Length', 'Area Latitude', 'Area Longitude'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3279 entries, 0 to 278
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Route           3279 non-null   object 
 1   Location        3279 non-null   object 
 2   URL             3279 non-null   object 
 3   Avg Stars       3279 non-null   float64
 4   Your Stars      3279 non-null   int64  
 5   Route Type      3279 non-null   object 
 6   Rating          3279 non-null   object 
 7   Pitches         3279 non-null   int64  
 8   Length          2834 non-null   float64
 9   Area Latitude   3279 non-null   float64
 10  Area Longitude  3279 non-null   float64
dtypes: float64(4), int64(2), object(5)
memory usage: 307.4+ KB


In [6]:
df.rename(columns={'Avg Stars': 'Stars_Avg'}, inplace=True)

## Vérification des données vide

In [7]:
df.columns.isna()

array([False, False, False, False, False, False, False, False, False,
       False, False])

In [8]:
df.set_index('Route', inplace=True)

## Création d'une liste de pays

In [9]:
countries_list = ["Albania", "Austria", "Belgium", "Bosnia-Herzegovina", "Bulgaria", "Croatia", "Cyprus", "Czech Republic", "Denmark", "Faroe Islands", "Finland", "France", "Germany", "Gibraltar", "Greece", "Hungary", "Iceland", "Ireland", "Italy", "Kosovo", "Lichtenstein", "Lithuania", "Luxembourg", "Malta", "Montenegro", "Netherlands", "North Macedonia", "Northern Ireland", "Norway", "Poland", "Portugal", "Romania", "Russia", "Serbia", "Slovakia", "Slovenia", "Spain", "Sweden", "Switzerland", "Ukraine","United Kingdom"]

In [10]:
def find_country(col):
    for country in countries_list:
        if country in col:
            return country

In [11]:
df['Country'] = df.Location.apply(find_country)

In [12]:
df.head()

Location  \
Route                                                                           
Cosmiques Arête             Vallée Blanche Basin > Mont Blanc Range > .Sno...   
Hörnligrat Route            Matterhorn > Zermatt area > Switzerland > Euro...   
Ferrata Brigata Tridentina  Via Ferrata > Passo Gardena area (Grodner Joch...   
Troutdale Pinnacle          Black Crag > Borrowdale Valley > The Lake Dist...   
Creagh Dhu Wall             Main Wall > Craig y Castell - Crag of the Cast...   

                                                                          URL  \
Route                                                                           
Cosmiques Arête             https://www.mountainproject.com/route/10749708...   
Hörnligrat Route            https://www.mountainproject.com/route/11132447...   
Ferrata Brigata Tridentina  https://www.mountainproject.com/route/11128278...   
Troutdale Pinnacle          https://www.mountainproject.com/route/10663200...   
Creagh Dhu Wall             https://www.mountainproject.com/route/10663944...   

                            Stars_Avg  Your Stars  \
Route                                               
Cosmiques Arête                   3.8          -1   
Hörnligrat Route                  3.9          -1   
Ferrata Brigata Tridentina        4.0          -1   
Troutdale Pinnacle                3.9          -1   
Creagh Dhu Wall                   3.9          -1   

                                                Route Type  \
Route                                                        
Cosmiques Arête             Trad, Mixed, Ice, Snow, Alpine   
Hörnligrat Route                              Trad, Alpine   
Ferrata Brigata Tridentina                      TR, Alpine   
Troutdale Pinnacle                                    Trad   
Creagh Dhu Wall                                       Trad   

                                          Rating  Pitches  Length  \
Route                                                               
Cosmiques Arête             5.6 AI2 M4 Mod. Snow        2  1000.0   
Hörnligrat Route                        5.4 PG13        1  3600.0   
Ferrata Brigata Tridentina                   4th        1  1600.0   
Troutdale Pinnacle                           5.6        6   350.0   
Creagh Dhu Wall                              5.6        3   180.0   

                            Area Latitude  Area Longitude         Country  
Route                                                                      
Cosmiques Arête                  45.87882         6.88555          France  
Hörnligrat Route                 45.97648         7.65921     Switzerland  
Ferrata Brigata Tridentina       46.51870        11.82130           Italy  
Troutdale Pinnacle               54.54424        -3.14018  United Kingdom  
Creagh Dhu Wall                  52.94180        -4.14660  United Kingdom

In [13]:
# Création d'une échelle de mesures avec 1 dictionnaire, on s'arrêtera au niveau 7 :

cotation_measure = {"5.4":1, "5.5":2, "5.6":3, "5.7":4, "5.8":5, "5.9":6, "5.10":7}


In [14]:
def find_cotation(col):
    for k,v in cotation_measure.items():
        if col[:3] in k:
            return v

In [15]:
df['cotation'] = df.Rating.apply(find_cotation)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3279 entries, Cosmiques Arête to El Woppo
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Location        3279 non-null   object 
 1   URL             3279 non-null   object 
 2   Stars_Avg       3279 non-null   float64
 3   Your Stars      3279 non-null   int64  
 4   Route Type      3279 non-null   object 
 5   Rating          3279 non-null   object 
 6   Pitches         3279 non-null   int64  
 7   Length          2834 non-null   float64
 8   Area Latitude   3279 non-null   float64
 9   Area Longitude  3279 non-null   float64
 10  Country         3279 non-null   object 
 11  cotation        3092 non-null   float64
dtypes: float64(5), int64(2), object(5)
memory usage: 333.0+ KB


## Suppression des données nulles

In [17]:
#Suppression des 

df = df[df['cotation'].notna()]

In [19]:
df = df[df['Length'].notna()]

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2655 entries, Cosmiques Arête to Nerfetiti
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Stars_Avg       2655 non-null   float64
 1   Route Type      2655 non-null   object 
 2   Pitches         2655 non-null   int64  
 3   Length          2655 non-null   float64
 4   Area Latitude   2655 non-null   float64
 5   Area Longitude  2655 non-null   float64
 6   Country         2655 non-null   object 
 7   cotation        2655 non-null   float64
 8   aid             2655 non-null   int64  
 9   alpine          2655 non-null   int64  
 10  boulder         2655 non-null   int64  
 11  ice             2655 non-null   int64  
 12  mixed           2655 non-null   int64  
 13  snow            2655 non-null   int64  
 14  sport           2655 non-null   int64  
 15  tr              2655 non-null   int64  
 16  trad            2655 non-null   int64  
dtypes: float64(5), int6

In [32]:
#Plus que 2655 enregistrements

## Supprimer les voies d'escalade sans aucun vote

In [21]:
list_index = df.query('Stars_Avg <= 0').index
len(list_index)

12

In [22]:
#Suppression des 12 enregistrements :

df.drop(list_index, axis=0, inplace=True)

In [23]:
#Vérification de la suppression

df.query('Stars_Avg <= 0')

Empty DataFrame
Columns: [Location, URL, Stars_Avg, Your Stars, Route Type, Rating, Pitches, Length, Area Latitude, Area Longitude, Country, cotation]
Index: []

## Supprimer les voies d'escalade qui ont emplacement inférieur strictement à 0 (impossible)

In [24]:
list_index_pitches = df.query('Pitches < 0').index
len(list_index_pitches)

2

In [25]:
#Suppression des 2 enregistrements :

df.drop(list_index_pitches, axis=0, inplace=True)

In [26]:
df.query('Pitches < 0')

Empty DataFrame
Columns: [Location, URL, Stars_Avg, Your Stars, Route Type, Rating, Pitches, Length, Area Latitude, Area Longitude, Country, cotation]
Index: []

## Conversion de la longueur de la voie d'escalade en mètres

In [27]:
df["Length"] = df.Length.apply(lambda x: x*0.3048)

## Création d'un "bag of words" pour analyser les données des types de voies d'escalade

### Les différents types de voies :

Aid = Type Via-Ferrata, des systèmes d'échelle permanente, des cordes prédisposées avec ligne de vie pour aider le grimpeur sur sa voie (Aid of etrier / aid-ladder)

Boulder = bloc, consiste à grimper des blocs de faible hauteur et sans assurage 

Sport = L'escalade sportive est un style d'escalade qui repose sur des ancrages permanents fixés sur le rocher

Trad = Le grimpeur part du bas de la voie et place ses protections au fur et à mesure (ancrages amovibles comme les coinceurs)

Ice = Escalade sur glacier

Snow = Voie avec de la neige

Mixed = L'escalade mixte ou le mixte est une méthode de progression en escalade qui utilise des crampons et des piolets pour grimper sur un environnement mixte composé de glace et de rocher

Alpine = Les rochers alpins diffèrent par leur qualité, leur comportement face aux intempéries et à la rupture, ainsi que leur forme. La connaissance des roches est donc nécessaire pour la sécurité de chaque grimpeur (les protections sont donc différentes et répondent à des normes supplémentaires)

TR = Top Rope, technique de corde supérieure pour pratiquer l'escalade en solo

In [28]:
text_data = np.array(df['Route Type'].values)
text_data

array(['Trad, Mixed, Ice, Snow, Alpine', 'Trad, Alpine', 'Trad', ...,
       'Trad', 'Trad', 'Trad'], dtype=object)

In [29]:
# Create the bag of words feature matrix
count = CountVectorizer()
bag_of_words = count.fit_transform(text_data)

# Show feature matrix
bag_of_words.toarray()

array([[0, 1, 0, ..., 0, 0, 1],
       [0, 1, 0, ..., 0, 0, 1],
       [0, 0, 0, ..., 0, 0, 1],
       ...,
       [0, 0, 0, ..., 0, 0, 1],
       [0, 0, 0, ..., 0, 0, 1],
       [0, 0, 0, ..., 0, 0, 1]], dtype=int64)

In [30]:
feature_names = count.get_feature_names()

In [31]:
df_types_one_hot_encoding = pd.DataFrame(bag_of_words.toarray(), columns=feature_names)
df_types_one_hot_encoding

aid  alpine  boulder  ice  mixed  snow  sport  tr  trad
0       0       1        0    1      1     1      0   0     1
1       0       1        0    0      0     0      0   0     1
2       0       0        0    0      0     0      0   0     1
3       0       0        0    0      0     0      0   0     1
4       0       0        0    0      0     0      0   0     1
...   ...     ...      ...  ...    ...   ...    ...  ..   ...
2650    0       0        0    0      0     0      1   0     0
2651    0       0        0    0      0     0      0   0     1
2652    0       0        0    0      0     0      0   0     1
2653    0       0        0    0      0     0      0   0     1
2654    0       0        0    0      0     0      0   0     1

[2655 rows x 9 columns]

## Concaténation du one-hot-encoding des types de voies d'escalade dans le dataframe initial

In [32]:
# Le one hot encoding ne fonctionnait pas sans le rapprochement des index numérotés, re-création des index pour concaténation :

df = df.reset_index()

In [33]:
df = pd.concat([df, df_types_one_hot_encoding], axis=1)

## Suppression des colonnes inutiles

In [35]:
df.drop(columns=['Location','Rating','Your Stars'], inplace=True)

In [36]:
df.drop(columns='URL', inplace=True)

In [37]:
df.set_index('Route', inplace=True)

## Premier aperçu résumé de la Data

In [38]:
df.describe()

Stars_Avg      Pitches       Length  Area Latitude  Area Longitude  \
count  2655.000000  2655.000000  2655.000000    2655.000000     2655.000000   
mean      3.204934     2.254614    66.090857      45.983433        8.384993   
std       0.897284     3.137149   146.192864       6.761746       10.070987   
min       1.000000     1.000000     1.828800      28.155560      -98.792090   
25%       3.000000     1.000000    15.240000      41.259500        0.074460   
50%       3.500000     1.000000    24.384000      46.163600        8.472100   
75%       4.000000     1.000000    39.624000      50.030600       12.466560   
max       4.000000    32.000000  4572.000000      68.298410       43.317800   

          cotation          aid       alpine      boulder          ice  \
count  2655.000000  2655.000000  2655.000000  2655.000000  2655.000000   
mean      5.359699     0.007156     0.071563     0.000377     0.001507   
std       2.076666     0.084308     0.257812     0.019407     0.038793   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%       3.000000     0.000000     0.000000     0.000000     0.000000   
50%       7.000000     0.000000     0.000000     0.000000     0.000000   
75%       7.000000     0.000000     0.000000     0.000000     0.000000   
max       7.000000     1.000000     1.000000     1.000000     1.000000   

             mixed         snow        sport           tr         trad  
count  2655.000000  2655.000000  2655.000000  2655.000000  2655.000000  
mean      0.001883     0.001883     0.683616     0.039925     0.312618  
std       0.043364     0.043364     0.465152     0.195819     0.463647  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     1.000000     0.000000     0.000000  
75%       0.000000     0.000000     1.000000     0.000000     1.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000

In [39]:
df.head()

Stars_Avg  \
Route                                             
Cosmiques Arête                             3.8   
Hörnligrat Route                            3.9   
Troutdale Pinnacle                          3.9   
Creagh Dhu Wall                             3.9   
Bosigran Ridge a.k.a. Commando Ridge        4.0   

                                                          Route Type  Pitches  \
Route                                                                           
Cosmiques Arête                       Trad, Mixed, Ice, Snow, Alpine        2   
Hörnligrat Route                                        Trad, Alpine        1   
Troutdale Pinnacle                                              Trad        6   
Creagh Dhu Wall                                                 Trad        3   
Bosigran Ridge a.k.a. Commando Ridge                            Trad        8   

                                        Length  Area Latitude  Area Longitude  \
Route                                                                           
Cosmiques Arête                        304.800       45.87882         6.88555   
Hörnligrat Route                      1097.280       45.97648         7.65921   
Troutdale Pinnacle                     106.680       54.54424        -3.14018   
Creagh Dhu Wall                         54.864       52.94180        -4.14660   
Bosigran Ridge a.k.a. Commando Ridge   152.400       50.17590        -5.62000   

                                             Country  cotation  aid  alpine  \
Route                                                                         
Cosmiques Arête                               France       3.0    0       1   
Hörnligrat Route                         Switzerland       1.0    0       1   
Troutdale Pinnacle                    United Kingdom       3.0    0       0   
Creagh Dhu Wall                       United Kingdom       3.0    0       0   
Bosigran Ridge a.k.a. Commando Ridge  United Kingdom       1.0    0       0   

                                      boulder  ice  mixed  snow  sport  tr  \
Route                                                                        
Cosmiques Arête                             0    1      1     1      0   0   
Hörnligrat Route                            0    0      0     0      0   0   
Troutdale Pinnacle                          0    0      0     0      0   0   
Creagh Dhu Wall                             0    0      0     0      0   0   
Bosigran Ridge a.k.a. Commando Ridge        0    0      0     0      0   0   

                                      trad  
Route                                       
Cosmiques Arête                          1  
Hörnligrat Route                         1  
Troutdale Pinnacle                       1  
Creagh Dhu Wall                          1  
Bosigran Ridge a.k.a. Commando Ridge     1

# Compter le nombre de fois qu'apparaît un type de voie d'escalade

In [40]:
df['alpine'].value_counts()

0    2465
1     190
Name: alpine, dtype: int64

# Suppression d'un outlier dans la hauteur + Vérification hauteur

In [48]:
df.query('Length > 4000').index

Index(['Peuterey Integrale'], dtype='object', name='Route')

In [53]:
df.drop('Peuterey Integrale', axis=0, inplace=True)

Vérification ok sur le site source de la possibilité qu'une voie d'escalade soit d'une longueur inférieure à 5 mètres, notamment un minimum à 1m82 (ex : Napes Needle https://www.mountainproject.com/route/106641638/napes-needle)

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2654 entries, Cosmiques Arête to Nerfetiti
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Stars_Avg       2654 non-null   float64
 1   Route Type      2654 non-null   object 
 2   Pitches         2654 non-null   int64  
 3   Length          2654 non-null   float64
 4   Area Latitude   2654 non-null   float64
 5   Area Longitude  2654 non-null   float64
 6   Country         2654 non-null   object 
 7   cotation        2654 non-null   float64
 8   aid             2654 non-null   int64  
 9   alpine          2654 non-null   int64  
 10  boulder         2654 non-null   int64  
 11  ice             2654 non-null   int64  
 12  mixed           2654 non-null   int64  
 13  snow            2654 non-null   int64  
 14  sport           2654 non-null   int64  
 15  tr              2654 non-null   int64  
 16  trad            2654 non-null   int64  
dtypes: float64(5), int6

Notre dataset comprend désormais 2654 voies d'escalade et 17 colonnes qui caractérisent ces voies.

## Export du Dataset

In [55]:
df.to_csv("C:/Users/amandine.gauberville/dataV2-labs/module-2/mini-project2/route-finder-rock-clean.csv", index=True)